### PR TITLE
Caching layer

### PR DESCRIPTION
**Pull request by ziazi**
This pull request provides the foundations of the offline mode in Versus App. More specifically, it includes the offline package with the following:
- A NetworkManager that provides a static method  --NetworkManager: : IsNetworkAvailable-- that given the context allows to check if the device has any kind of connection.
- A CachedPost class that abstracts the most important fields of Post stored in the local database used for the caching feature.
- A PostDatabase class, built using Room that represents the actual database that stores the Posts.
- A PostDAO that interacts with the database using SQL queries.
- A CacheManager which is a concrete implementation of DataBaseManager specific to posts and which provides an interface for caching, retrieving, inserting and deleting posts from the cache asynchronously.

This pull request also includes the necessary tests for  theses modifications that go through most possible cases.

### CODE REVIEW

**Review by DerouichAbdessalam — APPROVED** [review]
Good modularized code ! Great work on test coverage too

File: app/src/main/java/com/github/versus/offline/CacheManager.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public static CacheManager getCacheManager(Context context){
+        if(instance==null){
+            instance= new CacheManager(context);
+        }
+        return instance;
+    }
+    @Override
+    public Future<Boolean> insert(Post post) {
```
Comment (hamzaremmal) [inline app/src/main/java/com/github/versus/offline/CacheManager.java:33]: Maybe we can use [CompletableFuture](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletableFuture.html)

**Review by hamzaremmal — APPROVED** [review]
LGTM, all the tests seems good and we might use [CompletableFuture](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletableFuture.html) in the future :-)

File: app/src/main/java/com/github/versus/offline/CacheManager.java (hunk 1 above)
Comment (ziazi) [reply app/src/main/java/com/github/versus/offline/CacheManager.java:33]: Modifie that in DataBaseManager, the implementation returns a CompletableFuture anyway

**Comment by ziazi** [comment]
> LGTM, all the tests seems good and we might use [CompletableFuture](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletableFuture.html) in the future :-)

Modifie the interface and it's good 👍

**Pull request closed without merging**